rfc-120-csv-preview.md

```diff
@@ -0,0 +1,134 @@
+
+
+### Step 3
+
+Change Whitehall to use the new CSV Preview feture for newly published documents. We will need to use a different URL structure, since the current one is coupled to Whitehall.
+
+```
+# current URL example
+https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/856932/Management_information_-_further_education_and_skills_-_as_at_31_December_2019.csv/preview
+
+# proposed URL example
+https://www.gov.uk/government/remaining-parent-document-base-path/preview/asset-id
+```
+
+The `asset-id` should be the ID used by Asset Manager e.g. `597b098a759b743e0b759a96`, since it persists for the lifetime of the asset, even if the original file is replaced.
+
+
+### Step 4
+
+Change existing attachments (in Whitehall) to use the new CSV Preview. This will involve republishing all Whitehall documents that include a CSV preview link.
+
+Once this is done, we can remove some of the config for [the `assets` domain](https://github.com/alphagov/govuk-puppet/blob/51d7f7b648e8fde5aae3adfb774ec3bca6325bd8/modules/router/templates/assets_origin.conf.erb#L69-L74) that supported the legacy feature ([`gov.uk` routing](https://github.com/alphagov/government-frontend/blob/dcbc563030c470e6242d6432896d84feebb2138b/config/routes.rb#L14) remains). We can also delete the code for the legacy feature.
+
+> We may want to add redirects for the old preview URLs in Whitehall, if they continue to receive lots of traffic after all the 'View Online' links have updated.
+
+
+## Rejected proposals
+
+### Use a top-level prefix for the routing (e.g. `/preview/:base_path/:attachment_id`)
+
+This would involve registering a new `/preview` prefix route in [router][router], which would set the backend (rendering app) to Government Frontend for all paths beginning with that route.
+
+- This involves less effort than individually managing lower-level routes for each attachment preview. However, because we don't have any knowledge of these routes, we can't help the user if the URL they enter is invalid; this could happen when the asset is removed, or the parent document changes e.g. unpublished, redirected.
+
+- Using a top-level prefix for CSV Preview is at-odds with the behaviour of other top-level prefixes (`/info`, `/api/content`), since it only works for specific URLs, and doesn't apply to GOV.UK content in general.
+
+### Improve the feature in Whitehall (frontend)
+
+This would involve rewriting the existing CSV Preview feature so that it can work for documents published elsewhere, instead of being coupled to the attachment model in Whitehall.
+
+> We could make the feature work for attachments published by others apps, using the proposed URL structure, which gives us the parent document and the `asset-id` of attachment being previewed.
+
+- This would mean continued reliance of the Whitehall (frontend) codebase and the `/government` prefix route, which has been identified as [tech debt][government-prefix].
```

Comment:
I'm a bit confused by this, as your proposed URL structure is `https://www.gov.uk/government/remaining-parent-document-base-path/preview/asset-id` (i.e. it still has a `/government` prefix).

Reply:
Yes, that's right. The difference is that, with this rejected proposal, we'd be relying on the `/government` prefix to do the routing. With the preferred proposal, we'll be registering a more specific route for each content item, which will override the behaviour of the prefix - this is the trie mechanism of the Router in action.

Reply:
Thanks - I think this can be disambiguated in the ['URL proposal explanation' I suggest here](https://github.com/alphagov/govuk-rfcs/pull/122#discussion_r417836323).